### ISSUE TITLE
Scaffolder: Have a way to do integration tests for fetch:template

### ISSUE THREAD

**Issue by Pike** [backstage/backstage#6568]
The `fetch:template` action in the scaffolder doesn't support an out-of-action integration test path. With the cookiecutter templates, one could just run the CLI.

## Feature Suggestion

Offer the templating logic as an export of the scaffolder module, or even as CLI.

## Context

Authors of templates would like to run intergration tests, or also use their templates outside of the backstage creation pipeline.

**Comment by benjdlambert** [comment]
I can smell a `backstage-cli template --vars` or something here that could be useful.

**Comment by Pike** [comment]
I've toyed around with this a bit, but I wonder how to approach this.

How to set up a code path as close as possible to the actual scaffolder-backend? Does that effectively mean running the backend and talking to urls? It'd seem nice to just use a in-memory database for this, though?

Is there an existing code path I could mimic?

**Comment by eperdeme** [comment]
After reading up on the changes to the templater, we would really appreciate a slim CLI tool similar to what cookiecutter provided for this.

Our use case is similar to the above where we need CI on occasion to generate results from templates so we can create disposable infrastructure to run tests against, prior to the branch being merged and then made available as "self service" in backstage for internal customers.

**Comment by freben** [comment]
I guess it depends how far one wants to take this.

The `createFetchTemplateAction` function is exported from the package, but as you have noticed it uses the `UrlReader` and `ScmIntegrations` facilities for starters. Those can indeed be initialised from config if one can get at the config, but if anything related to actual production remotes is referenced by the template, those will need secrets to be present as well, in order to be able to perform their work.

In an integration test scenario within js code, you could of course just pass mocks/fakes to those parameters.

In a CLI scenario, maybe we could load the config (and accept `--config` flags similar to what the build etc commands do). Not sure how to best approach the secrets part then.

And after that, the template itself will of course have a number of required parameters that will need to be filled in.

I realise that this wasn't a very helpful comment, I just wanted to brain dump a little bit and note that we haven't forgotten about the idea entirely :)

**Comment by Pike** [comment]
I've done a local test of an unrelated templater patch, and here's what I learned:

I can't get local files to work unless I actually link them up in the config. There's a bunch of CORS until I finally hit the URLReader only knowing how to load a tree from SCM integrations.
I tweaked `TaskWorker.ts` to not remove the workspacePath, and ensured I had cookiecutter installed to not end up in a docker container, and then I could inspect the temp dir for the result.

If one was to test a template that was built to publish to GH or so, that would obviously be trickier still.

But with a running backend on a spoofed config, it's almost feasible ;-)

**Comment by benjdlambert** [comment]
So I'm seeing two things in this discussion here, It looks like theres a request to have the same functionality that we had with `cookiecutter`, a standalone CLI that you can run on top of a template directory with some values (this bit is what I think is important short term), and another way to be able to run a `kind: Template` thing for integration tests.

Would that be a fair assumption?

I'm thinking that we should really just extract the logic that's inside the action handler and wrap that up in a CLI and provide some sort of CLI command that people can run on top of a skeleton directory with some json value input or arguments and then validate that the output it correct?

**Comment by eperdeme** [comment]
So from my perspective my main use case is the CLI one you wrote above. 

We found https://github.com/Tradeshift/nunjucks-cli/tree/nunjucks a few days ago which goes most of the way to make it work by feeding in a data.json file and outputting the rendered template. This allowed us to render the output of templates to do lint/integration tests on via CI pipelines. However there's no guarantees it'll continue to always work as backstage evolves.

**Comment by benjdlambert** [comment]
Yep indeed, we want to make a CLI which can wrap up the `values` part and any additional options that are passed to it.

Maybe we could look at building a CLI wrapper around `actions` that can just run one action in a workspace, so we could have a little bit of both.

Thinking for the `fetch:template` usage you have something like the following:

```js
// input.json
{
 "url": "./template",
 "values": { "something": "My Test Value" },
 "cookiecutterCompat": true
}
```
And and then you could do 

```sh
$ backstage-cli actions run fetch:template --input input.json
```

Also feel however, that there's probably quite a lot of work in getting something like that up and running, so maybe short term we can just expose the `nunjucks` templating logic as seperate CLI command.

**Comment by msamad** [comment]
My two cents:
Being a fan of out of the process component tests I am thinking of two approaches for our app.

1, Run the whole backstage app (frontend + backend)  and use mocked systems instead of real ones. There are a bit of hardcoded endpoints in there e.g. bitbucket.org endpoint which we are using but it can be solved by using containers and creating host aliases to point to local mock container. Of coarse, updating upstream backstage project to pickup all endpoints from config would be ideal though to avoid host alias trick and we can just provide a config to point to mock container.
Regardless, the app can be tested end-to-end by using mocked systems instead of real ones which is great because this will allow us to test the behavior without relying on external systems and will create a fast feedback loop for dev.

2. It is similar to first approach but instead of testing the whole backstage app (frontend + backend). Why not test them as two separate components, this may makes things easier to test, especially for the frontend testing. The beauty of using APIs come out here. The frontend doesn't really care who the backend is, as long as the API call comes back with correct response from the backend - backend is real or mock, doesn't matter.
This allows to put more focus on frontend if needed to make sure it behaves exactly the way it is suppose to given the right data.
And the backend APIs can be tested independently, all we need to make sure is that given the right input, they produce the right output. And of coarse, the same problem of redirecting requests to mock systems instead of real systems is still there but the solution in first approach will work here too.

**Comment by Pike** [comment]
I realize I didn't say what I actually wanted to test, and which hat I'm wearing doing so. The hat I'm wearing is that of a template author and maintainer, and I'd like to know if the template(s) I did work.

I'd like to be able to manually inspect the output, and also run linters or included tests. I want to do so for a number of different combinations of answers. I might be stacking different templates, too.

For a maintainer, I'd also like to see if any package updates in the template break the template itself. Say, you install Django, or create-react-app, do things continue to work? This would be run on a cron schedule, probably.

**Comment by msamad** [comment]
Sorry, I may have dumped my comments above without much context. It was after reading this comment
> How to set up a code path as close as possible to the actual scaffolder-backend? Does that effectively mean running the backend and talking to urls? It'd seem nice to just use a in-memory database for this, though?

If it is possible to call the api that backend-scaffolder provides, it should produce the output from a template, including performing the required fetch and publish tasks. e.g. let's say it fetches the template from github, creates an output and then publishes to github. Now instead of actual github, if there is a mock endpoint from where it fetches and then publishes, we could grab the output from mocked endpoint (could also grab it from output directory) and then verify whatever we see fit e.g manual checks, linters, automated tests to verify output actually is buildable/runnable.  

This could also open up the possibility of testing the template and all of it's actions whenever the app code changes, plus the added advantage of verifying API calls to external systems e.g. github.

**Comment by Pike** [comment]
Side note, https://github.com/backstage/backstage/discussions/7116 is related.

**Comment by jrusso1020** [comment]
Hi all, (full disclosure only skimmed through this issue/still reading through it)

We are also interested in this.

For context, we have other teams owning almost all of our templates and when they are testing their templates they just want to see the template step output resembles what they expect it to. Previously we told engineers to use cookiecutter locally for this and even had scripts to help automate this

We aren't necessarily trying to recreate the whole scaffolder, we still find testing it in Backstage end to end locally handles this fine. The main issue for us is with the move away from cookiecutter to templating in JS with nunjunks, we'd like a similar CLI/method to test our template with some dummy values gives us the expected output

Although testing all the actions via a CLI could be nice, I think a good first step would just be testing the nunjunks template step via a CLI

**Comment by timja** [comment]
Not stale

**Comment by sabinasiddiqi** [comment]
Our team would also like this feature ✋

**Comment by cnmcavoy** [comment]
As a new user of backstage, I'm quite confused at the state of `fetch:template`, given there seems to be supported way to test rendering templates locally with any sort of input, the `nunjucks-cli` referenced earlier is now a dead link, and I haven't found any issues that explains why `cookiecutter` was abandoned as the engine, given the project seems alive. The docs do not explain any of this.

Can someone give some context and guidance as to whether backstage is supporting templating at all? My initial impression is that I am better off writing a custom action to invoke a well-known templating engine rather than try to decipher any of this.

**Comment by cnmcavoy** [comment]
Those reasons make sense to me, I was surprised I could not find them written down or [mentioned in the docs](https://backstage.io/docs/features/software-templates/migrating-from-v1beta2-to-v1beta3). It's good to hear that cookiecutter is still supported, and still planned on being supported, as a plugin - the only examples I have been able to find well documented are for cookiecutter. Thanks for the explanation.

**Comment by fedyiv** [comment]
I am confused with the template format used in backstage.  It was said above that nunjucks is used for rendering templated.
But in [nunjucks docs](https://mozilla.github.io/nunjucks/templating.html) it is mentioned that following syntax is used for variables: `{{ variable }}` while in backstage templates it is `${{ variable }}`.

Where is the difference comes from? Because of this we could not use nunjucks-cli for local rendering of the templates

**Comment by seanpoulter-foci** [comment]
Thank you all, with extra thanks to @freben for mentioning `createFetchTemplateAction`. If anyone else finds this thread looking for ways to test the output of the `fetch:template` action, [this example](https://github.com/PHACDataHub/sci-portal/blob/d47d99f5858c1e4dfa4dc80572dc9def93a8467b/backstage/packages/backend/src/plugins/scaffolder/templates/project-create.test.ts) might help.

**Comment by Stannislav** [comment]
> I am confused with the template format used in backstage. It was said above that nunjucks is used for rendering templated. But in [nunjucks docs](https://mozilla.github.io/nunjucks/templating.html) it is mentioned that following syntax is used for variables: `{{ variable }}` while in backstage templates it is `${{ variable }}`.
> 
> Where is the difference comes from? Because of this we could not use nunjucks-cli for local rendering of the templates

@fedyiv I wondered about the same and found [this](https://backstage.io/docs/features/software-templates/migrating-from-v1beta2-to-v1beta3/#--instead-of--).

TLDR: `{{ }}` interferes with YAML syntax in `template.yaml`, while `${{ }}` doesn't.
